api/core/storage/slack/slack_api_client.py

```diff
@@ -140,3 +162,86 @@ async def get_channel_info(self, channel_id: str):
             operation_name="get slack channel info",
         )
         return self.ChannelInfo.model_validate(parsed["channel"])
+
+    def _should_process_webhook_event(self, webhook_event: SlackWebhookEvent) -> bool:
+        bot_id: str | None = None
+
+        # Filter out message that do not contain "@WorkflowAI Bot"
+        if len(webhook_event.authorizations) > 0:
+            # TODO: use an env var to store the bot id
+            bot_id = webhook_event.authorizations[0].user_id
+
+            if webhook_event.event and webhook_event.event.text and bot_id not in webhook_event.event.text:
+                _logger.info(
+                    "The message is not addressed to the bot, skipping",
+                    extra={"event": webhook_event.event.text},
+                )
+                return False
+
+        # Filter messages send by the bot itself
+        if webhook_event.is_bot_triggered():
+            _logger.info("Skipping bot triggered event", extra={"event": webhook_event})
+            return False
+
+        # Filter out non-message events (will be handled later, maybe)
+        if webhook_event.event.type != "message":
+            _logger.info("Skipping non-message event", extra={"event": webhook_event.event})
+            return False
+
+        return True
+
+    async def handle_webhook(self, raw_payload: dict[str, Any]) -> None:
+        try:
+            # Parse the payload as a SlackWebhookEvent
+            webhook_event = SlackWebhookEvent(**raw_payload)
+            if not self._should_process_webhook_event(webhook_event=webhook_event):
+                return
+
+            # We need to implement a Redis lock mechanism because events are sometimes sent 2-3 times by Slack
+            client_msg_id = getattr(webhook_event.event, "client_msg_id", None)
+            if not client_msg_id:
+                _logger.warning(
+                    "No client_msg_id found in event, processing without lock",
+                    extra={"event": webhook_event.event},
+                )
+            else:
+                lock_key = f"slack:webhook:lock:{client_msg_id}"
+                try:
+                    async with redis_lock(lock_key, expire_seconds=60):
+                        # Process the message only if we obtained the lock
+                        await self._process_webhook_message(webhook_event)
+                        return
+                except LockAcquisitionError:
+                    # This message is already being processed or was processed recently
+                    _logger.info("Skipping duplicate webhook event", extra={"client_msg_id": client_msg_id})
+                    return
+
+            await self._process_webhook_message(webhook_event)
+
+        except Exception as e:
+            _logger.exception("Failed to handle Slack webhook", extra={"payload": raw_payload, "error": e})
+
+    async def _process_webhook_message(self, webhook_event: SlackWebhookEvent) -> None:
```

Comment:
Careful, this is in the wrong layer. The slack client should only know about slack and not about our internal processes.

The easiest way to respect layer boundaries would be to have `handle_webhook` return a validated webhook event or None.
And move the below logic to a class method in the customer service that you can call in the router directly

Reply:
yes yes the business logic should go elsewhere. I just rushed to push the code.<br><br>But I can adjust.

Reply:
done in https://github.com/WorkflowAI/WorkflowAI/pull/176/commits/e69a20d7b23e909eaf380e9a569deadf2fcce8b5